proposals/target-typed-new.md

```diff
@@ -31,37 +31,29 @@ Instantiate an object without spelling out the type.
 private readonly static object s_syncObj = new();
 ```
 
-## Detailed design
+## Specification
 [design]: #detailed-design
 
-The *object_creation_expression* syntax would be modified to make the *type* optional when parentheses are present. This is required to address the ambiguity with *anonymous_object_creation_expression*.
+A new syntactic form, *target_typed_new* of the *object_creation_expression* is accepted in which the *type* is optional.
+
 ```antlr
 object_creation_expression
-    : 'new' type? '(' argument_list? ')' object_or_collection_initializer?
+    : 'new' type '(' argument_list? ')' object_or_collection_initializer?
     | 'new' type object_or_collection_initializer
+    | target_typed_new
+    ;
+target_typed_new
+    : 'new' '(' argument_list? ')' object_or_collection_initializer?
     ;
 ```
 
-A target-typed `new` is convertible to any type. As a result, it does not contribute to overload resolution. This is mainly to avoid unpredictable breaking changes.
-
-The argument list and the initializer expressions will be bound after the type is determined.
-
-The type of the expression would be inferred from the target-type which would be required to be one of the following:
-
-- **Any struct type** (including tuple types)
-- **Any reference type** (including delegate types)
-- **Any type parameter** with a constructor or a `struct` constraint
+A *target_typed_new* expression does not have a type. However, there is a new *object creation conversion* that is an implicit conversion from expression, that exists from a *target_typed_new* to every type. It is an error for this conversion to be applied unless the type being converted to is
+- a `struct` type; or
+- a `class` type; or
+- a `delegate` type; or
+- a type parameter.
```

Comment:
The previous revision had some additional details on type parameter case, which I think should be kept:

> **Any type parameter** with a constructor or a `struct` constraint

Reply:
Why?  It will be a semantic error to apply the conversion.  I don't think we should condition the existence of the conversion on constraints.

---
In reply to: [397972206](https://github.com/dotnet/csharplang/pull/3310#discussion_r397972206) [](ancestors = 397972206)